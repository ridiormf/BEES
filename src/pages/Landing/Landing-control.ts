import React from 'react';

export const useLandingControl = () => {
  const [fullName, setFullName] = React.useState<string>('');
  const [majority, setMajority] = React.useState<boolean>(false);

  const onChangeFullName = (
    event: React.ChangeEvent<HTMLInputElement>,
  ): void => {
    setFullName(event.target.value);
  };

  const onChangeMajority = (
    event: React.ChangeEvent<HTMLInputElement>,
  ): void => {
    setMajority(event.target.checked);
  };

  return {
    state: {
      fullName,
      majority,
    },
    methods: {
      onChangeFullName,
      onChangeMajority,
    },
  };
};
