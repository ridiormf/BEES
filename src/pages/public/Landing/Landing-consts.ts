export const LANDING = {
  TEST_ID: {
    CONTAINER: 'LANDING_CONTAINER',
    FORM: 'LANDING_FORM',
    FORM_DESCRIPTION: 'LANDING_FORM_DESCRIPTION',
    FORM_HINT: 'LANDING_FORM_HINT',
    INPUT_TEXT: 'LANDING_INPUT_TEXT',
    LABEL_CHECKBOX: 'LANDING_LABEL_CHECKBOX',
    INPUT_CHECKBOX: 'LANDING_INPUT_CHECKBOX',
    SUBMIT_BUTTON: 'LANDING_SUBMIT_BUTTON',
    LANDING_IMAGE: 'LANDING_LANDING_IMAGE',
  },
};
