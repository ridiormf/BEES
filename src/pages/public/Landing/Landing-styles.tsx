import { CSS } from '@stitches/react';
import theme from '../../../theme';

const Container = theme.styled('div', {
  alignSelf: 'stretch',
  flexGrow: 1,
  backgroundColor: '$primaryBackground',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  paddingHorizontal: 20,
});

const FormContainer = theme.styled('form', {
  width: 300,
  display: 'flex',
  alignItems: 'start',
  flexDirection: 'column',
  zIndex: 1,
});

const textFieldCSS: CSS = {
  marginTop: 6,
  marginBottom: 16,
  textTransform: 'capitalize',
};

const checkboxCSS: CSS = {
  alignSelf: 'center',
};

const buttonCSS: CSS = {
  marginTop: 22,
  alignSelf: 'center',
};

const imageCSS: CSS = {
  position: 'absolute',
  left: 28,
  bottom: 16,
  maxWidth: '100%',
  '@sm': {
    left: '50%',
    transform: 'translateX(-50%)',
    width: '40%',
  },
};

export const landingStyles = {
  styled: {
    Container,
    FormContainer,
  },
  css: {
    textField: textFieldCSS,
    checkbox: checkboxCSS,
    button: buttonCSS,
    image: imageCSS,
  },
};
