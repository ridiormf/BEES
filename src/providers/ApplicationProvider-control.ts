import React from 'react';
import { ApplicationProviderControl, User } from './ApplicationProvider-types';

export const ApplicationContext =
  React.createContext<ApplicationProviderControl>({
    state: {},
    methods: {},
  });

export const useApplicationProviderControl = (): ApplicationProviderControl => {
  const [user, setUser] = React.useState<User>();

  const saveUser = (userToSave: User) => {
    setUser(userToSave);
  };

  const clearUser = () => {
    setUser(undefined);
  };

  return {
    state: {
      user,
    },
    methods: {
      saveUser,
      clearUser,
    },
  };
};

export const useApplicationContext = (): ApplicationProviderControl => {
  const applicationContext = React.useContext(ApplicationContext);
  return applicationContext;
};
