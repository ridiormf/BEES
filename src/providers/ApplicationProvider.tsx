import {
  ApplicationContext,
  useApplicationProviderControl,
} from './ApplicationProvider-control';
import { ApplicationProviderProps } from './ApplicationProvider-types';

const ApplicationProvider: React.FC<ApplicationProviderProps> = ({
  children,
}) => {
  const control = useApplicationProviderControl();
  return (
    <ApplicationContext.Provider value={control}>
      {children}
    </ApplicationContext.Provider>
  );
};

export default ApplicationProvider;
