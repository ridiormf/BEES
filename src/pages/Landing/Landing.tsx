import React from 'react';

import { useLandingControl } from './Landing-control';
import { landingStyles } from './Landing-styles';
import { LandingProps } from './Landing-types';

import { Button } from '../../components/Button';
import { Checkbox } from '../../components/Checkbox';
import { Input } from '../../components/Input';
import { Text } from '../../components/Text';
import { Image } from '../../components/Image';

import BeesImage from '../../assets/png/bees.png';

const {
  styled: { Container, FormContainer },
  css,
} = landingStyles;

const Landing: React.FC<LandingProps> = () => {
  const {
    state: { fullName },
    methods: { onChangeFullName },
  } = useLandingControl();
  return (
    <Container>
      <FormContainer>
        <Text>Please, enter your full name below</Text>
        <Text>Only alphabetical characters are accepted</Text>
        <Input
          onChange={onChangeFullName}
          value={fullName}
          css={css.textField}
        />
        <Checkbox
          id="fullName"
          label="Are you older than 18 years old?"
          css={css.checkbox}
        />
        <Button css={css.button}>Enter</Button>
      </FormContainer>
      <Image src={BeesImage} css={css.image} />
    </Container>
  );
};

export default Landing;
