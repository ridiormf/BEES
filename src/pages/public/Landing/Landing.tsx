import React from 'react';

import { useLandingControl } from './Landing-control';
import { landingStyles } from './Landing-styles';
import { LandingProps } from './Landing-types';

import Button from '../../../components/Button';
import Checkbox from '../../../components/Checkbox';
import Input from '../../../components/Input';
import Text from '../../../components/Text';
import Image from '../../../components/Image';

import BeesImage from '../../../assets/png/bees.png';
import { LANDING } from './Landing-consts';

const {
  styled: { Container, FormContainer },
  css,
} = landingStyles;

const Landing: React.FC<LandingProps> = () => {
  const {
    state: { fullName, majority, validForm },
    methods: { onChangeFullName, onBlurFullName, onChangeMajority, onSaveForm },
  } = useLandingControl();
  return (
    <Container data-testid={LANDING.TEST_ID.CONTAINER}>
      <FormContainer data-testid={LANDING.TEST_ID.FORM} onSubmit={onSaveForm}>
        <Text data-testid={LANDING.TEST_ID.FORM_DESCRIPTION}>
          Please, enter your full name below
        </Text>
        <Text data-testid={LANDING.TEST_ID.FORM_HINT} css={css.secondText}>
          Only alphabetical characters are accepted
        </Text>
        <Input
          data-testid={LANDING.TEST_ID.INPUT_TEXT}
          onChange={onChangeFullName}
          value={fullName}
          placeholder="Full name"
          onBlur={onBlurFullName}
          css={css.textField}
        />
        <Checkbox
          data-testid={LANDING.TEST_ID.LABEL_CHECKBOX}
          id="fullName"
          label="Are you older than 18 years old?"
          inputProps={{
            checked: majority,
            onChange: onChangeMajority,
            'data-testid': LANDING.TEST_ID.INPUT_CHECKBOX,
          }}
          css={css.checkbox}
        />
        <Button
          data-testid={LANDING.TEST_ID.SUBMIT_BUTTON}
          css={css.button}
          disabled={!validForm}
          type="submit"
        >
          Enter
        </Button>
      </FormContainer>
      <Image
        data-testid={LANDING.TEST_ID.LANDING_IMAGE}
        src={BeesImage}
        css={css.image}
        alt="Bee image"
      />
    </Container>
  );
};

export default Landing;
