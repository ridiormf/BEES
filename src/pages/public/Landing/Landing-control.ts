import React from 'react';
import { useApplicationContext } from '../../../providers/ApplicationProvider';
import { checkFullName } from '../../../utils/validations';

export const useLandingControl = () => {
  const [fullName, setFullName] = React.useState<string>('');
  const [majority, setMajority] = React.useState<boolean>(false);

  const [validForm, setValidForm] = React.useState<boolean>(false);

  const applicationContext = useApplicationContext();

  const onChangeFullName = (
    event: React.ChangeEvent<HTMLInputElement>,
  ): void => {
    if (validForm) {
      setValidForm(false);
    }
    setFullName(event.target.value);
  };

  const onChangeMajority = (
    event: React.ChangeEvent<HTMLInputElement>,
  ): void => {
    setValidForm(event.target.checked && checkFullName(fullName.trim()));
    setMajority(event.target.checked);
  };

  const onBlurFullName = () => {
    setValidForm(majority && checkFullName(fullName.trim()));
  };

  const onSaveForm = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    applicationContext.openFullLoading!();
    applicationContext.saveUser!({
      fullName: fullName.trim(),
      majority,
    });
  };

  return {
    state: {
      fullName,
      majority,
      validForm,
    },
    methods: {
      onChangeFullName,
      onChangeMajority,
      onBlurFullName,
      onSaveForm,
    },
  };
};
