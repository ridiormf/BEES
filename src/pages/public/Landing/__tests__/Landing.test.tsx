import { fireEvent, render } from '@testing-library/react';
import { Wrapper } from '../../../../__tests__/utils';

import * as applicationProviderMethods from '../../../../providers/ApplicationProvider/ApplicationProvider-control';

import Landing from '../Landing';
import { LANDING } from '../Landing-consts';

const mockUseApplicationContext = ({
  saveUser = jest.fn(),
  openFullLoading = jest.fn(),
  showNewFeedback = jest.fn(),
}) => {
  return jest
    .spyOn(applicationProviderMethods, 'useApplicationContext')
    .mockImplementation(() => ({
      saveUser,
      openFullLoading,
      showNewFeedback,
    }));
};

jest.useFakeTimers();

describe('pages ->', () => {
  describe('public ->', () => {
    describe('Landing ->', () => {
      describe('snapshots ->', () => {
        it('Should match snapshot', () => {
          const { container } = render(<Landing />, { wrapper: Wrapper });
          expect(container).toMatchSnapshot();
        });
      });
      describe('controls ->', () => {
        it('Should not submit form if checkbox is unchecked and full name valid', () => {
          const openFullLoading = jest.fn();
          mockUseApplicationContext({ openFullLoading });
          const { getByTestId } = render(<Landing />, { wrapper: Wrapper });
          const textInput = getByTestId(LANDING.TEST_ID.INPUT_TEXT);
          const button = getByTestId(LANDING.TEST_ID.SUBMIT_BUTTON);

          fireEvent.change(textInput, { target: { value: 'Full name' } });
          fireEvent.blur(textInput);
          fireEvent.click(button);
          expect(openFullLoading).not.toBeCalled();
        });
        it('Should not submit form if checkbox is checked and full name invalid', () => {
          const openFullLoading = jest.fn();
          mockUseApplicationContext({ openFullLoading });
          const { getByTestId } = render(<Landing />, { wrapper: Wrapper });
          const checkbox = getByTestId(LANDING.TEST_ID.INPUT_CHECKBOX);
          const input = getByTestId(LANDING.TEST_ID.INPUT_TEXT);
          const button = getByTestId(LANDING.TEST_ID.SUBMIT_BUTTON);

          fireEvent.change(input, { target: { value: 'full n' } });
          fireEvent.click(checkbox);
          fireEvent.click(button);

          expect(openFullLoading).not.toBeCalled();
        });
        it('Should not submit form if checkbox is toggle to unchecked and full name invalid', () => {
          const openFullLoading = jest.fn();
          mockUseApplicationContext({ openFullLoading });
          const { getByTestId } = render(<Landing />, { wrapper: Wrapper });
          const checkbox = getByTestId(LANDING.TEST_ID.INPUT_CHECKBOX);
          const button = getByTestId(LANDING.TEST_ID.SUBMIT_BUTTON);
          const input = getByTestId(LANDING.TEST_ID.INPUT_TEXT);
          fireEvent.change(input, { target: { value: 'full name' } });
          fireEvent.click(checkbox);
          fireEvent.click(checkbox);
          fireEvent.click(button);
          expect(openFullLoading).not.toBeCalled();
        });

        it('Should not submit form if checkbox is checked and full name toggle from valid to invalid', () => {
          const openFullLoading = jest.fn();
          mockUseApplicationContext({ openFullLoading });
          const { getByTestId } = render(<Landing />, { wrapper: Wrapper });
          const checkbox = getByTestId(LANDING.TEST_ID.INPUT_CHECKBOX);
          const button = getByTestId(LANDING.TEST_ID.SUBMIT_BUTTON);
          const input = getByTestId(LANDING.TEST_ID.INPUT_TEXT);
          fireEvent.change(input, { target: { value: 'full name' } });
          fireEvent.click(checkbox);
          fireEvent.change(input, { target: { value: 'full n' } });
          fireEvent.blur(input);
          fireEvent.click(button);
          expect(openFullLoading).not.toBeCalled();
        });

        it('Should submit form if checkbox is checked and full name valid', () => {
          const openFullLoading = jest.fn();
          const showNewFeedback = jest.fn();
          const saveUser = jest.fn();
          mockUseApplicationContext({
            openFullLoading,
            showNewFeedback,
            saveUser,
          });
          const { getByTestId } = render(<Landing />, { wrapper: Wrapper });
          const checkbox = getByTestId(LANDING.TEST_ID.INPUT_CHECKBOX);
          const button = getByTestId(LANDING.TEST_ID.SUBMIT_BUTTON);
          const input = getByTestId(LANDING.TEST_ID.INPUT_TEXT);
          fireEvent.change(input, { target: { value: 'full name' } });
          fireEvent.click(checkbox);
          fireEvent.click(button);
          expect(openFullLoading).toBeCalled();
          expect(showNewFeedback).toBeCalled();
          expect(saveUser).not.toBeCalled();
          jest.runAllTimers();
          expect(saveUser).toBeCalled();
        });
      });
    });
  });
});
