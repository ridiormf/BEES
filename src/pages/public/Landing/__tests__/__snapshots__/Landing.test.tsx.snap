// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages -> public -> Landing -> snapshots -> Should match snapshot 1`] = `
<div>
  <div
    class="c-jjBNkD"
    data-testid="LANDING_CONTAINER"
  >
    <form
      class="c-jbNAsZ"
      data-testid="LANDING_FORM"
    >
      <p
        class="c-caxBfD"
        data-testid="LANDING_FORM_DESCRIPTION"
      >
        Please, enter your full name below
      </p>
      <p
        class="c-caxBfD c-caxBfD-ikujBtM-css"
        data-testid="LANDING_FORM_HINT"
      >
        Only alphabetical characters are accepted
      </p>
      <input
        class="c-jyoysL c-jyoysL-ihCWOuT-css"
        data-testid="LANDING_INPUT_TEXT"
        placeholder="Full name"
        value=""
      />
      <label
        class="c-dKuRBS c-dKuRBS-iiyfMXp-css"
        data-testid="LANDING_LABEL_CHECKBOX"
        for="fullName"
      >
        <input
          class="c-inUAee"
          data-testid="LANDING_INPUT_CHECKBOX"
          id="fullName"
          type="checkbox"
        />
         
        Are you older than 18 years old?
      </label>
      <button
        class="c-gHLNgH c-gHLNgH-ifjAMLC-css"
        data-testid="LANDING_SUBMIT_BUTTON"
        disabled=""
        type="submit"
      >
        Enter
      </button>
    </form>
    <img
      alt="Bee image"
      class="PJLV PJLV-ifuKsSQ-css"
      data-testid="LANDING_LANDING_IMAGE"
      src="bees.png"
    />
  </div>
</div>
`;
